### PR TITLE
WantAuthnRequestsSigned IDPSSODescriptor is not configurable in resident IDP. #7323 

### PR DESCRIPTION
**Pull request by SiluniPathirana**
…ptor is not configurable in resident IDP. #7323

### Proposed changes in this pull request
https://github.com/wso2/product-is/issues/7323

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]
Add the following property s to the identity.xml file reside in the <IS_HOME>/repository/conf/identity folder under SSOService

```
 <SAML2AuthnRequestsSigningEnabled>false</SAML2AuthnRequestsSigningEnabled>
```
-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/IdentityConstants.java [hunk 1]
```diff
@@ -249,6 +249,10 @@ private IdentityConstants() {
         public static final String SAML_SESSION_NOT_ON_OR_AFTER_PERIOD = "SSOService.SAMLSessionNotOnOrAfterPeriod";
         public static final String SAML_METADATA_SIGNING_ENABLED = "SSOService.SAMLMetadataSigningEnabled";
         public static final String SAML_ECP_URL = "SSOService.SAMLECPEndpoint";
+        public static final String SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED="SSOService.WantAuthnRequestsSigned";
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/IdentityConstants.java]: limit the length to 120.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/org/wso2/carbon/idp/mgt/ui/i18n/Resources.properties [hunk 2]
```diff
@@ -433,6 +433,11 @@ script.default.seq.placeholder=Please add a default authentication sequence conf
 saml.metadata.validity.period=Metadata Validity Period
 saml.metadata.validity.period.help=Set the SAML metadata validity period in minutes
 saml.metadata.signing.enabled=Enable metadata signing
+saml.metadata.wantAuthnRequests.signing.enabled= Enable AuthnRequestsSigned
+saml.metadata.name.id.format=NameIDFormat
+saml.metadata.attribute.profile= Attribute Profile
+
+
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/org/wso2/carbon/idp/mgt/ui/i18n/Resources.properties]: remove the redundant empty line.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 3]
```diff
@@ -111,6 +111,27 @@
         }
         property.setValue(samlMetadataSigningEnabled);
         propertyList.add(property);
+
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED);
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: limit the length to 120.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 4]
```diff
@@ -111,6 +111,27 @@
...
+        String WantAuthnRequestsSignedEnabled = "false";
+        if (StringUtils.containsIgnoreCase(request.getParameter("WantAuthnRequestsSigned"), "on")) {
+                    WantAuthnRequestsSignedEnabled = "true";
+                }
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: format this.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 5]
```diff
@@ -108,6 +108,11 @@
     String samlMetadataValidityPeriod = null;
     boolean samlMetadataSigningEnabled = false;
     String samlMetadataSigningEnabledChecked = "";
+    boolean wantAuthnRequestsSignedEnabled = false;
+    String wantAuthnRequestsSignedChecked = "";
+    String samlNameIDFormat=null;
+    String attributeProfile=null;
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: format the lines.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 6]
```diff
@@ -139,6 +144,20 @@
...
+
+                         }
+
+             samlNameIDFormat=IdPManagementUIUtil.getProperty(properties,
+                                 IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_NAME_ID_FORMAT).getValue();
+
+             attributeProfile=IdPManagementUIUtil.getProperty(properties,
+                                              IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_ATTRIBUTE_PROFILE).getValue();
+
+
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: format the lines.

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java [hunk 7]
```diff
@@ -180,6 +183,15 @@ public IdentityProvider getResidentIdP(String tenantDomain)
             samlECPUrl = IdentityUtil.getServerURL(IdentityConstants.ServerConfig.SAMLSSO, true, true);
         }
 
+        WantAuthnRequestsSigned = IdentityUtil.getServerURL(IdentityConstants.ServerConfig.SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED, true, true)
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java]: limit the length to 120.

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java [hunk 8]
```diff
@@ -180,6 +183,15 @@ public IdentityProvider getResidentIdP(String tenantDomain)
...
+                + IdPManagementUtil.getTenantParameter();
+
+        samlNameIDFormat=IdentityUtil.getServerURL(IdentityConstants.ServerConfig.SAML_METADATA_NAME_ID_FORMAT, true, true)
+                + IdPManagementUtil.getTenantParameter();
+
+        attributeProfile=IdentityUtil.getServerURL(IdentityConstants.ServerConfig.SAML_METADATA_ATTRIBUTE_PROFILE, true, true)
```
Comment (thanujalk) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java]: limit the length to 120.

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml [hunk 9]
```diff
@@ -577,6 +577,8 @@
         <!-- Request validity period in minutes-->
         <SAML2AuthenticationRequestValidityPeriod>5</SAML2AuthenticationRequestValidityPeriod>
         <SAMLSPCertificateExpiryValidationEnabled>false</SAMLSPCertificateExpiryValidationEnabled>
+        <NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:transient</NameIDFormat>
+        <AttributeProfile>http://sample.com</AttributeProfile>
```
Comment (thanujalk) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml]: Do we need default value for this? Can't we comment this?

Comment (thanujalk) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml]: same changes need to go to identity.xml.j2.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java [hunk 10]
```diff
@@ -174,6 +174,10 @@ private OpenID() {
             public static final String SAML_METADATA_VALIDITY_PERIOD_DEFAULT = "60";
             public static final String SAML_METADATA_SIGNING_ENABLED = "SAMLMetadataSigningEnabled";
             public static final String SAML_METADATA_SIGNING_ENABLED_DEFAULT = "false";
+            public static final String SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED="WantAuthnRequestsSigned";
+            public static final String SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_DEFAULT="false";
+            public static final String SAML_METADATA_NAME_ID_FORMAT="NameIDFormat";
+            public static final String SAML_METADATA_ATTRIBUTE_PROFILE="AttributeProfile";
```
Comment (sachiniWettasinghe) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: formatting issue

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 11]
```diff
@@ -111,6 +111,27 @@
...
+        property.setValue(WantAuthnRequestsSignedEnabled);
+        propertyList.add(property);
+
+
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: remove redundant line

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 12]
```diff
@@ -111,6 +111,27 @@
         }
         property.setValue(samlMetadataSigningEnabled);
         propertyList.add(property);
+
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED);
+        String WantAuthnRequestsSignedEnabled = "false";
+        if (StringUtils.containsIgnoreCase(request.getParameter("WantAuthnRequestsSigned"), "on")) {
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: Can't we use here the constant defined in IdentityApplicationConstants class for "WantAuthnRequestsSigned"?

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 13]
```diff
@@ -111,6 +111,27 @@
         }
         property.setValue(samlMetadataSigningEnabled);
         propertyList.add(property);
+
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED);
+        String WantAuthnRequestsSignedEnabled = "false";
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: Can't we assign SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_DEFAULT constant defined in IdentityApplicationConstants class for this?

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 14]
```diff
@@ -111,6 +111,27 @@
...
+
+
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_NAME_ID_FORMAT);
+        property.setValue(request.getParameter("NameIDFormat"));
+        propertyList.add(property);
+
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_ATTRIBUTE_PROFILE);
+        property.setValue(request.getParameter("AttributeProfile"));
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: Let's use the previously defined constants for "NameIDFormat" and "AttributeProfile"

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 15]
```diff
@@ -139,6 +144,20 @@
             if (samlMetadataSigningEnabled) {
                 samlMetadataSigningEnabledChecked = "checked=\'checked\'";
             }
+             wantAuthnRequestsSignedEnabled = Boolean.parseBoolean(IdPManagementUIUtil.getProperty(properties,
+                                 IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_WANT_AUTHN_REQUESTS_SIGNED_ENABLED).getValue());
+                         if (wantAuthnRequestsSignedEnabled) {
+                             wantAuthnRequestsSignedChecked = "checked=\'checked\'";
+
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: remove the redundant line fix formatting issues

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java [hunk 16]
```diff
@@ -470,6 +482,37 @@ public IdentityProvider getResidentIdP(String tenantDomain)
...
+                IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_ATTRIBUTE_PROFILE);
+        if (samlAttributeProfileProperty == null) {
+            samlAttributeProfileProperty = new Property();
+            samlAttributeProfileProperty.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_ATTRIBUTE_PROFILE);
+            samlSSOUrlProperty.setValue(attributeProfile);
+        }
+        propertiesList.add(samlAttributeProfileProperty);
+
+
+
```
Comment (sachiniWettasinghe) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java]: remove redundant lines

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java [hunk 17]
```diff
@@ -174,7 +174,8 @@ private OpenID() {
             public static final String SAML_METADATA_VALIDITY_PERIOD_DEFAULT = "60";
             public static final String SAML_METADATA_SIGNING_ENABLED = "SAMLMetadataSigningEnabled";
             public static final String SAML_METADATA_SIGNING_ENABLED_DEFAULT = "false";
-
+            public static final String SAML_METADATA_AUTHN_REQUESTS_SIGNING_ENABLED ="samlAuthnRequestsSigningEnabled";
```
Comment (emswbandara) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: formatting issue

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java [hunk 18]
```diff
@@ -174,7 +174,8 @@ private OpenID() {
...
+            public static final String SAML_METADATA_AUTHN_REQUESTS_SIGNING_DEFAULT ="false";
```
Comment (emswbandara) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: formatting issue

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml [hunk 19]
```diff
@@ -469,6 +469,7 @@
         <SAMLDefaultAssertionEncryptionAlgorithmURI>http://www.w3.org/2001/04/xmlenc#aes256-cbc</SAMLDefaultAssertionEncryptionAlgorithmURI>
         <SAMLDefaultKeyEncryptionAlgorithmURI>http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p</SAMLDefaultKeyEncryptionAlgorithmURI>
         <SLOHostNameVerificationEnabled>true</SLOHostNameVerificationEnabled>
+        <SAMLAuthnRequestsSignedEnabled>false</SAMLAuthnRequestsSignedEnabled>
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/resources/repository/conf/identity/identity.xml]: Make the config as "SAML2AuthnRequestsSigningEnabled" similar to https://github.com/wso2-support/carbon-identity-framework/pull/933/files#diff-58e3089af8f5a12e8b92c629111ca555R490

File: components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/IdentityConstants.java [hunk 20]
```diff
@@ -249,6 +249,8 @@ private IdentityConstants() {
         public static final String SAML_SESSION_NOT_ON_OR_AFTER_PERIOD = "SSOService.SAMLSessionNotOnOrAfterPeriod";
         public static final String SAML_METADATA_SIGNING_ENABLED = "SSOService.SAMLMetadataSigningEnabled";
         public static final String SAML_ECP_URL = "SSOService.SAMLECPEndpoint";
+        public static final String SAML_METADATA_AUTHN_REQUESTS_SIGNING_ENABLED ="SSOService" +
```
Comment (emswbandara) [inline components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/IdentityConstants.java]: formatting issue

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/org/wso2/carbon/idp/mgt/ui/i18n/Resources.properties [hunk 21]
```diff
@@ -432,9 +432,11 @@ title.add.default.seq.select.mode=Select Mode
 script.default.seq.placeholder=Please add a default authentication sequence configuration in XML
 saml.metadata.validity.period=Metadata Validity Period
 saml.metadata.validity.period.help=Set the SAML metadata validity period in minutes
-saml.metadata.signing.enabled=Enable metadata signing
+saml.metadata.signing.enabled=Enable metadata Signing
```
Comment (emswbandara) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/org/wso2/carbon/idp/mgt/ui/i18n/Resources.properties]: let's make M upper case in the resource value. it should be 'Enable Metadata Signing'

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java [hunk 22]
```diff
@@ -1107,8 +1139,7 @@ private void setExpressionNodeList(Node node, List<ExpressionNode> expression)
                                         message);
                     }
                 }
-            }
-            expression.add((ExpressionNode) node);
+            }expression.add((ExpressionNode) node);
```
Comment (emswbandara) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java]: this change seems unnecessary. Let's revert it

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp [hunk 23]
```diff
@@ -111,6 +111,14 @@
         }
         property.setValue(samlMetadataSigningEnabled);
         propertyList.add(property);
+        property = new Property();
+        property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.SAML_METADATA_AUTHN_REQUESTS_SIGNING_ENABLED);
+        String samlAuthnRequestsSigningEnabled = "false";
+        if (StringUtils.containsIgnoreCase(request.getParameter("samlAuthnRequestsSigningEnabled"), "on")) {
```
Comment (Abilashini) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: Can we use the defined constant(SAML_METADATA_AUTHN_REQUESTS_SIGNING_ENABLED) here

Comment (SiluniPathirana) [reply components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local-ajaxprocessor.jsp]: other request params are obtained using the same way. Hence you followed the same for consistency

**Pull request merged**